### PR TITLE
Namespaced Reflect wrapper class

### PR DESCRIPTION
**Pull request by deepakrkris**
* Scope reflect context to a namespace

* Provide wrapper methods for reflect api

cc @raymondfeng

### CODE REVIEW

File: packages/context/src/reflect.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {inject, Constructor} from '..';
+import 'reflect-metadata';
+
+/*
+ * wrapper to handle reflect api
+ */
+export class ReflectWrapper {
```
Comment (bajtos) [inline packages/context/src/reflect.ts]: I find this name not descriptive enough, it's not clear what is being wrapped.

I am proposing something like `NamespacedReflect` or `ScopedReflect`.

Comment (deepakrkris) [inline packages/context/src/reflect.ts]: agreed

**Comment by deepakrkris** [comment]
cc @bajtos @raymondfeng

File: packages/context/src/reflect.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+   * @param namespace : namespace to bind this reflect context
+   */
+  constructor(private namespace?: string) {}
+
+  /**
+   * lookup metadata from a target object and its prototype chain
+   */
+  getMetadata(metadataKey: string, target: Function, propertyKey?: string | symbol): Object {
+    // prefix namespace, if provided, to the metadata key
+    metadataKey = this.namespace? this.namespace + ':' + metadataKey : metadataKey;
```
Comment (bajtos) [inline packages/context/src/reflect.ts]: The implementation of prefixing algorithm is duplicated too much to my taste. Have you considered extracting it into a shared method?

```ts
export class NamespacedReflect {
  // ...
  getMetadata(metadataKey: string, target: Function, propertyKey?: string | symbol): Object {
    metadataKey = this.buildKey(metadataKey);
    // ...
  }

  buildKey(metadataKey: string) {
    return this.namespace? this.namespace + ':' + metadataKey : metadataKey;
  }
}
```

Comment (bajtos) [inline packages/context/src/reflect.ts]: Or perhaps `this.addNamespace(metadataKey)` may be a better method name.

File: packages/context/src/reflect.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    if (propertyKey)
+      return Reflect.getMetadata(metadataKey, target, propertyKey);
+    return Reflect.getMetadata(metadataKey, target);
```
Comment (superkhau) [inline packages/context/src/reflect.ts]:
```
return propertyKey ?
  Reflect.getMetadata(metadataKey, target, propertyKey) :
  Reflect.getMetadata(metadataKey, target)
```

**Comment by raymondfeng** [comment]
@deepakrkris I had to commit this code into `feature/persistence` - https://github.com/strongloop/loopback-next/pull/312/commits/7ee74dbaab0227295437deaf18651b73afce1df1. 

The key is to make sure we have a singleton for all references as `reflect-metadata` use globals. Adding `reflect-metadata` as dependencies to multiple packages exhibits strange behaviors.

Here is my proposal to further enhance your PR to implement all methods from `Reflect` and export `NamespacedReflect` class and a singleton `export const Reflector = new NamespacedReflect('loopback')`.

**Comment by raymondfeng** [comment]
@deepakrkris I have improved the PR based on the latest master. Please review.

File: packages/context/src/reflect.ts [hunk 4]
```diff
@@ -3,18 +3,153 @@
...
+  }
+
+  getMetadataKeys(target: Object, propertyKey?: string | symbol): string[] {
+    let keys: string[];
+    if (propertyKey) {
+      keys = Reflect.getMetadataKeys(target, propertyKey);
+    } else {
+      keys = Reflect.getMetadataKeys(target);
+    }
+    const metaKeys = [];
```
Comment (deepakrkris) [inline packages/context/src/reflect.ts]: nitpick, can we change this to `let metaKeys;` and assign empty array [] only on required condition

File: packages/context/src/reflect.ts [hunk 5]
```diff
@@ -3,18 +3,153 @@
...
+  }
+
+  getOwnMetadataKeys(target: Object, propertyKey?: string | symbol): string[] {
+    let keys: string[];
+    if (propertyKey) {
+      keys = Reflect.getOwnMetadataKeys(target, propertyKey);
+    } else {
+      keys = Reflect.getOwnMetadataKeys(target);
+    }
+    const metaKeys = [];
```
Comment (deepakrkris) [inline packages/context/src/reflect.ts]: same as above

File: packages/context/test/unit/reflect.ts [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+
+      // get the defined metadata using the namespaced reflectContext
+      let metadata = reflectContext.getMetadata('key', BaseClass);
+      expect(metadata).to.be.equal(metadataValue);
+
+      metadata = reflectContext.getOwnMetadata('key', BaseClass);
+      expect(metadata).to.be.equal(metadataValue);
+
+      metadata = reflectContext.getOwnMetadata('key', SubClass);
+      // expect(metadata).to.be.undefined();
```
Comment (deepakrkris) [inline packages/context/test/unit/reflect.ts]: do we need this test for own metadata ? it may not be applicable here

Comment (raymondfeng) [inline packages/context/test/unit/reflect.ts]: This one is interesting. I was expecting line 78 returns `undefined`.

File: packages/context/test/unit/reflect.ts [hunk 7]
```diff
@@ -0,0 +1,728 @@
...
+      expect(result).to.be.false();
+
+      result = reflectContext.hasOwnMetadata('key2', BaseClass);
+      expect(result).to.be.true();
+    });
+
+  });
+
+  describe('default Reflector', () => {
+    beforeEach(givenReflectContext);
```
Comment (deepakrkris) [inline packages/context/test/unit/reflect.ts]: do we need to add all test cases for the default Reflector ? wont it be enough to check the type of the default reflector ?

**Comment by deepakrkris** [comment]
LGTM

**Pull request merged**